File: utilities/nonvisualstudio/elixir/Spec/Math.Spec.ex
# IN POWERSHELL
# iex.bat - REPL
# elixir c:\Users\Thomas\Documents\GitHub\practice\utilities\nonvisualstudio\elixir\Spec\Math.Spec.ex

Code.require_file "../Implementation/Math.ex", __DIR__

ExUnit.start

defmodule MathTests do
  use ExUnit.Case, async: true

  test "gcd(990,210) should return 30." do
    assert Math.gcd(990, 210) === 30
  end

  test "gcd(210,990) should return 30." do
    assert Math.gcd(210, 990) === 30
  end

  test "gcd(952,624) should return 8." do
    assert Math.gcd(952, 624) === 8
  end

  test "gcd(624,952) should return 8." do
    assert Math.gcd(624, 952) === 8
  end

  test "lcm(65,50) should return 650." do
    assert Math.lcm(65,50) === 650
  end

  test "lcm(10,4) should return 20." do
    assert Math.lcm(10,4) === 20
  end

  test "pow 100 10 should return 100_000_000_000_000_000_000." do
    assert Math.pow(100,10) === 100_000_000_000_000_000_000
  end

  test "pow 5 5 should return 3_125." do
    assert Math.pow(5,5) === 3_125
  end

  test "powMod 4 13 497 should return 445." do
    assert Math.powMod(4,13,497) === 445
  end

  test "powMod 46_565 240_437_235 961_748_941 should return 533_506_838." do
    assert Math.powMod(46_565, 240_437_235, 961_748_941) === 533_506_838
  end

  test "powMod 46_565 0 961_748_941 should return 1." do
    assert Math.powMod(46_565, 0, 961_748_941) === 1
  end

end

File: utilities/nonvisualstudio/haskell/Spec/Math.Spec.hs
{-
ghci c:\Users\Thomas\Documents\GitHub\practice\utilities\nonvisualstudio\haskell\Spec\Math.Spec.hs c:\Users\Thomas\Documents\GitHub\practice\utilities\nonvisualstudio\haskell\Implementation\Math.hs
-}
-- :r - reload
-- :q - quit
-- :set +s for times

module MathTests where

  import Test.QuickCheck
  import Test.HUnit
  import System.IO


  import Math

  powModTests = TestList
    [
      TestCase $ assertEqual "powMod 4 13 497 should return 445." 445 (powMod 4 13 497),
      TestCase $ assertEqual "powMod 46565 0 961748941 should return 1." 1 (powMod 46565 0 961748941),
      TestCase $ assertEqual "powMod 46565 240437235 961748941 should return 533506838." 533506838 (powMod 46565 240437235 961748941),
      TestCase $ assertEqual "powMod 45621 419039293 301230039 should return 6285651." 6285651 (powMod 45621 419039293 301230039)
    ]



  testCases = TestList
    [
      powModTests
    ]

  qc_powMod_exponentOne :: Integer -> Integer -> Property
  qc_powMod_exponentOne base modulus =
    modulus > 0 ==>
      powMod base 1 modulus == base `mod` modulus

  qc_powMod :: Integer -> Integer -> Integer -> Property
  qc_powMod base theExponent modulus =
    modulus > 0 && theExponent > 1 ==>
      powMod base theExponent modulus == base ^ theExponent `mod` modulus

  tests = runTestTT testCases
    >> quickCheck qc_powMod_exponentOne
    >> quickCheck qc_powMod
